horovod/run/elastic/driver.py

```diff
@@ -0,0 +1,233 @@
+        self._discovery_thread.start()
+
+    def start(self, np, create_worker_fn):
+        self._create_worker_fn = create_worker_fn
+        self._activate_hosts(np)
+
+    def resume(self):
+        self._activate_hosts(self._min_np)
+
+    def stop(self):
+        self._shutdown.set()
+        self._discovery_thread.join()
+
+    def finished(self):
+        return self._shutdown.is_set()
+
+    def get_results(self):
+        return self._results.get_results()
+
+    def register_worker_server(self, host, slot, addresses, secret_key):
+        self._worker_clients[(host, slot)] = WorkerNotificationClient(
+            addresses, secret_key, self._verbose)
+
+    def record_ready(self, host, slot):
+        self._worker_registry.record_ready(host, slot)
+
+    def world_size(self):
+        return self._world_size
+
+    def local_size(self, host):
+        return len(self._host_assignments[host])
+
+    def get_slot_info(self, host, slot):
+        return self._host_assignments[host][slot] if self.has_rank_assignment(host, slot) \
+            else hosts.INVALID_SLOT_INFO
+
+    def has_rank_assignment(self, host, slot):
+        if self._discovered_hosts.is_blacklisted(host):
+            return False
+        return host in self._host_assignments and len(self._host_assignments[host]) > slot
+
+    def get_available_hosts(self):
+        return self._discovered_hosts.get_available_hosts()
+
+    def wait_for_available_hosts(self, min_np, max_np=None):
+        tmout = timeout.Timeout(
+            self._start_timeout,
+            message='Timed out waiting for {{activity}}. Please check that you have '
+                    'enough resources to run at least {min_np} Horovod processes.'.format(min_np=min_np))
+
+        self._wait_hosts_cond.acquire()
+        try:
+            while not self._has_available_slots(self._discovered_hosts.count_available_slots(), min_np, max_np):
+                self._wait_hosts_cond.wait(tmout.remaining())
+                tmout.check_time_out_for('minimum number of hosts to become available')
+        finally:
+            self._wait_hosts_cond.release()
+
+    def _has_available_slots(self, slots, min_np, max_np):
+        return slots >= min_np and (max_np is None or slots <= max_np)
+
+    def _activate_hosts(self, min_np):
+        logging.info('wait for available hosts: {}'.format(min_np))
+        self.wait_for_available_hosts(min_np)
+        new_assigned_hosts = self._update_assigned_hosts()
+        self._worker_registry.reset(self.world_size())
+        for host in new_assigned_hosts:
+            logging.info('start worker processes: {}'.format(host))
+            self._start_worker_processes(host)
+
+    def _discover_hosts(self):
+        while not self._shutdown.is_set():
+            self._wait_hosts_cond.acquire()
+            try:
+                if self._discovered_hosts.update_available_hosts():
+                    self._notify_workers_host_changes()
```

Comment:
Good point, I've added a precondition check to `_notify_workers_host_changes()` to early return if (1) no assigned hosts have been removed and (2) there are no added hosts that can be assigned (`max_np` slots already assigned or no available hosts were added).

We do already notify workers if an assigned host is removed from the available hosts, at which point it can be gracefully removed without needing to fail on allreduce.  The worker will kill itself after rendezvous when it received -1 for its rank.

You brought up a good optimization that can be made in a follow-up PR.  Currently, if a host is removed then added back before a reset event, then the reset will still occur (workers will raise an exception).  But ideally, the driver should send a notification to cancel the reset in such a case.

Reply:
Please create an issue for that optimization, or we will forget about it. You could label all those issues with a `elastic` label.

Reply:
Here we go: https://github.com/horovod/horovod/issues/1872